core/src/main/scala/org/apache/spark/deploy/rest/StandaloneRestClient.scala

```diff
@@ -62,45 +67,94 @@ private[deploy] class StandaloneRestClient extends Logging {
+            throw new SubmitRestConnectionException(
+              "No master is available for createSubmission.", new Throwable(""))
+          }
+      }
     }
     response
   }
 
   /** Request that the server kill the specified submission. */
-  def killSubmission(master: String, submissionId: String): SubmitRestProtocolResponse = {
+  def killSubmission(submissionId: String): SubmitRestProtocolResponse = {
     logInfo(s"Submitting a request to kill submission $submissionId in $master.")
-    validateMaster(master)
-    val response = post(getKillUrl(master, submissionId))
-    response match {
-      case k: KillSubmissionResponse => handleRestResponse(k)
-      case unexpected => handleUnexpectedRestResponse(unexpected)
+    var suc: Boolean = false
+    var response: SubmitRestProtocolResponse = null
+    for (m <- masters if !suc) {
+      validateMaster(m)
+      val url = getKillUrl(m, submissionId)
+      try {
+        response = post(url)
+        response match {
+          case k: KillSubmissionResponse =>
+            if (!k.message.contains("Can only")) {
+              handleRestResponse(k)
+              suc = true
+            }
+          case unexpected =>
+            handleUnexpectedRestResponse(unexpected)
+        }
+      } catch {
+        case e @ (_: SubmitRestConnectionException | _: ConnectException) =>
+          if(handleSubmitRestConnectionException(m)) {
+            throw new SubmitRestConnectionException(
+              "No master is available for killSubmission.", new Throwable(""))
+          }
+      }
     }
     response
   }
 
   /** Request the status of a submission from the server. */
   def requestSubmissionStatus(
-      master: String,
       submissionId: String,
       quiet: Boolean = false): SubmitRestProtocolResponse = {
     logInfo(s"Submitting a request for the status of submission $submissionId in $master.")
-    validateMaster(master)
-    val response = get(getStatusUrl(master, submissionId))
-    response match {
-      case s: SubmissionStatusResponse => if (!quiet) { handleRestResponse(s) }
-      case unexpected => handleUnexpectedRestResponse(unexpected)
+    var suc: Boolean = false
+    var response: SubmitRestProtocolResponse = null
+    for (m <- masters) {
+      validateMaster(m)
+      val url = getStatusUrl(m, submissionId)
+      try {
+        response = get(url)
+        response match {
+          case s: SubmissionStatusResponse if s.success =>
+            if (!quiet) {
+              handleRestResponse(s)
+            }
+            suc = true
+          case unexpected =>
+            handleUnexpectedRestResponse(unexpected)
+        }
+      } catch {
+        case e @ (_: SubmitRestConnectionException | _: ConnectException) =>
+          if(handleSubmitRestConnectionException(m)) {
+            throw new SubmitRestConnectionException(
+              "No master is available for requestSubmissionStatus.", new Throwable(""))
+          }
+      }
```

Comment:
This chunk of code seems to be for the most part duplicated. It would be good to factor it out.